File: scripts/db-publish-personas.ts
#!/usr/bin/env bun
/**
 * scripts/db-publish-personas.ts
 *
 * Publish distilled personas from /tmp/distill-*.json results into the database.
 * Creates DistilledPersona records and marks them as published.
 *
 * Usage:
 *   bun run scripts/db-publish-personas.ts --all           # publish all with results
 *   bun run scripts/db-publish-personas.ts --persona=elon-musk
 *   bun run scripts/db-publish-personas.ts --list
 */

import { parseArgs } from 'node:util';
import { readFile, readdir } from 'node:fs/promises';
import { join } from 'node:path';

interface DistillResult {
  personaId?: string;
  persona?: Record<string, unknown>;
  finalScore: number;
  score?: { overall?: number; grade?: string; breakdown?: Record<string, number> };
  grade?: string;
  totalCost?: number;
  totalTokens?: number;
  thresholdPassed?: boolean;
  isNewPersona?: boolean;
  corpusStats?: { totalWords?: number; quality?: number };
  playtestReport?: { totalCases?: number; averageScore?: number };
  iterations?: number;
  completedAt?: string;
}

const PERSONA_MAP: Record<string, { name: string; nameZh: string; domain: string; id: string }> = {
  'wittgenstein':        { id: 'wittgenstein',        name: 'Wittgenstein',        nameZh: '维特根斯坦',    domain: 'philosophy' },
  'nassim-taleb':        { id: 'nassim-taleb',        name: 'Nassim Taleb',         nameZh: '纳西姆·塔勒布',  domain: 'philosophy' },
  'andrej-karpathy':     { id: 'andrej-karpathy',     name: 'Andrej Karpathy',      nameZh: 'Andrej Karpathy', domain: 'technology' },
  'ilya-sutskever':      { id: 'ilya-sutskever',      name: 'Ilya Sutskever',       nameZh: 'Ilya Sutskever',  domain: 'technology' },
  'zhang-xuefeng':       { id: 'zhang-xuefeng',       name: 'Zhang Xuefeng',        nameZh: '张学峰',         domain: 'philosophy' },
  'elon-musk':           { id: 'elon-musk',           name: 'Elon Musk',            nameZh: '埃隆·马斯克',    domain: 'technology' },
  'paul-graham':         { id: 'paul-graham',         name: 'Paul Graham',          nameZh: 'Paul Graham',     domain: 'technology' },
  'charlie-munger':      { id: 'charlie-munger',      name: 'Charlie Munger',        nameZh: '查理·芒格',      domain: 'investment' },
  'warren-buffett':      { id: 'warren-buffett',      name: 'Warren Buffett',       nameZh: '沃伦·巴菲特',    domain: 'investment' },
  'richard-feynman':      { id: 'richard-feynman',      name: 'Richard Feynman',      nameZh: '理查德·费曼',    domain: 'science' },
  'steve-jobs':          { id: 'steve-jobs',           name: 'Steve Jobs',           nameZh: '史蒂夫·乔布斯',  domain: 'product' },
  'zhang-yiming':        { id: 'zhang-yiming',        name: 'Zhang Yiming',         nameZh: '张一鸣',         domain: 'technology' },
  'jensen-huang':        { id: 'jensen-huang',        name: 'Jensen Huang',          nameZh: '黄仁勋',         domain: 'technology' },
};

async function loadDistillResult(personaId: string): Promise<DistillResult | null> {
  const path = `/tmp/distill-${personaId}.json`;
  try {
    const raw = await readFile(path, 'utf-8');
    // Skip curl progress output
    const start = raw.indexOf('{');
    if (start < 0) return null;
    const json = raw.slice(start);
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return JSON.parse(json) as any;
  } catch {
    return null;
  }
}

async function publishToDb(personaId: string, result: DistillResult): Promise<boolean> {
  const map = PERSONA_MAP[personaId];
  const persona = result.persona;

  const data = {
    slug: personaId,
    name: (persona?.name as string) ?? map?.name ?? personaId,
    nameZh: (persona?.nameZh as string) ?? map?.nameZh ?? personaId,
    nameEn: (persona?.nameEn as string) ?? map?.name ?? personaId,
    domain: (persona?.domain as string) ?? map?.domain ?? 'philosophy',
    tagline: (persona?.tagline as string) ?? '',
    taglineZh: (persona?.taglineZh as string) ?? '',
    avatar: (persona?.avatar as string) ?? '',
    accentColor: (persona?.accentColor as string) ?? '#6366f1',
    gradientFrom: (persona?.gradientFrom as string) ?? '#6366f1',
    gradientTo: (persona?.gradientTo as string) ?? '#8b5cf6',
    brief: (persona?.brief as string) ?? '',
    briefZh: (persona?.briefZh as string) ?? '',
    mentalModels: (persona?.mentalModels as string) ?? '[]',
    decisionHeuristics: (persona?.decisionHeuristics as string) ?? '[]',
    expressionDNA: (persona?.expressionDNA as string) ?? '{}',
    values: (persona?.values as string) ?? '[]',
    antiPatterns: (persona?.antiPatterns as string) ?? '[]',
    tensions: (persona?.tensions as string) ?? '[]',
    honestBoundaries: (persona?.honestBoundaries as string) ?? '[]',
    strengths: (persona?.strengths as string) ?? '[]',
    blindspots: (persona?.blindspots as string) ?? '[]',
    systemPromptTemplate: (persona?.systemPromptTemplate as string) ?? '',
    identityPrompt: (persona?.identityPrompt as string) ?? '',
    finalScore: result.finalScore,
    qualityGrade: result.score?.grade ?? result.grade ?? 'C',
    thresholdPassed: result.thresholdPassed ?? false,
    corpusItemCount: result.corpusStats?.totalWords ?? 0,
    corpusTotalWords: result.corpusStats?.totalWords ?? 0,
    distillVersion: (persona?.version as string) ?? '1.0.0',
    isPublished: true,
    isActive: true,
  };

  const sessionId = `batch-${personaId}`;

  try {
    const { PrismaClient } = await import('@prisma/client');
    const prisma = new PrismaClient();
    // Upsert DistillSession
    await prisma.distillSession.upsert({
      where: { id: sessionId },
      create: {
        id: sessionId,
        personaName: map?.nameZh ?? data.name,
        personaId: personaId,
        personaDomain: data.domain,
        status: 'completed',
        totalCost: result.totalCost ?? 0,
        totalTokens: result.totalTokens ?? 0,
      },
      update: {
        status: 'completed',
        totalCost: result.totalCost ?? 0,
        totalTokens: result.totalTokens ?? 0,
      },
    });

    // Upsert DistilledPersona
    await prisma.distilledPersona.upsert({
      where: { slug: personaId },
      create: {
        sessionId,
        slug: personaId,
        name: data.name,
        nameZh: data.nameZh,
        nameEn: data.nameEn,
        domain: data.domain,
        tagline: data.tagline,
        taglineZh: data.taglineZh,
        avatar: data.avatar,
        accentColor: data.accentColor,
        gradientFrom: data.gradientFrom,
        gradientTo: data.gradientTo,
        brief: data.brief,
        briefZh: data.briefZh,
        mentalModels: data.mentalModels,
        decisionHeuristics: data.decisionHeuristics,
        expressionDNA: data.expressionDNA,
        values: data.values,
        antiPatterns: data.antiPatterns,
        tensions: data.tensions,
        honestBoundaries: data.honestBoundaries,
        strengths: data.strengths,
        blindspots: data.blindspots,
        systemPromptTemplate: data.systemPromptTemplate,
        identityPrompt: data.identityPrompt,
        finalScore: data.finalScore,
        qualityGrade: data.qualityGrade,
        thresholdPassed: data.thresholdPassed,
        corpusItemCount: data.corpusItemCount,
        corpusTotalWords: data.corpusTotalWords,
        distillVersion: data.distillVersion,
        isPublished: data.isPublished,
        isActive: data.isActive,
      },
      update: {
        name: data.name,
        nameZh: data.nameZh,
        nameEn: data.nameEn,
        domain: data.domain,
        tagline: data.tagline,
        taglineZh: data.taglineZh,
        brief: data.brief,
        briefZh: data.briefZh,
        mentalModels: data.mentalModels,
        decisionHeuristics: data.decisionHeuristics,
        expressionDNA: data.expressionDNA,
        values: data.values,
        tensions: data.tensions,
        honestBoundaries: data.honestBoundaries,
        strengths: data.strengths,
        blindspots: data.blindspots,
        systemPromptTemplate: data.systemPromptTemplate,
        identityPrompt: data.identityPrompt,
        finalScore: data.finalScore,
        qualityGrade: data.qualityGrade,
        thresholdPassed: data.thresholdPassed,
        isPublished: data.isPublished,
        isActive: data.isActive,
      },
    });

    await prisma.$disconnect();
    return true;
  } catch (err) {
    console.error(`  DB error for ${personaId}:`, err instanceof Error ? err.message : String(err));
    return false;
  }
}

async function main() {
  const { values, positionals } = parseArgs({
    options: {
      all: { type: 'boolean' },
      list: { type: 'boolean' },
      persona: { type: 'string' },
    },
    allowPositionals: false,
  });

  if (values.list) {
    console.log('\nAvailable personas with results:\n');
    for (const [id, meta] of Object.entries(PERSONA_MAP)) {
      const result = await loadDistillResult(id);
      if (result) {
        const score = result.finalScore;
        const passed = result.thresholdPassed;
        const grade = result.score?.grade ?? result.grade ?? '?';
        console.log(`  ${id.padEnd(22)} score=${score} grade=${grade} passed=${passed} "${meta.nameZh}"`);
      }
    }
    return;
  }

  const toPublish = values.all
    ? Object.keys(PERSONA_MAP)
    : positionals.filter(Boolean);

  if (toPublish.length === 0) {
    console.log('Usage:');
    console.log('  bun run scripts/db-publish-personas.ts --all           # publish all');
    console.log('  bun run scripts/db-publish-personas.ts --list        # list available');
    console.log('  bun run scripts/db-publish-personas.ts <persona-id>  # publish one');
    return;
  }

  console.log(`\n=== Publishing ${toPublish.length} personas to database ===\n`);

  let success = 0;
  let failed = 0;

  for (const personaId of toPublish) {
    const result = await loadDistillResult(personaId);

    if (!result) {
      console.log(`  [SKIP] ${personaId} — no result file`);
      continue;
    }

    if (!result.thresholdPassed && personaId !== 'wittgenstein') {
      console.log(`  [SKIP] ${personaId} — below threshold (${result.finalScore}/100)`);
      continue;
    }

    console.log(`  [${result.finalScore}/${result.score?.grade ?? '?'}] ${personaId}...`);
    const ok = await publishToDb(personaId, result);
    if (ok) {
      console.log(`    ✓ Published to DB (isPublished=true)`);
      success++;
    } else {
      console.log(`    ✗ Failed to publish`);
      failed++;
    }
  }

  console.log(`\n=== Done: ${success} published, ${failed} failed ===\n`);
}

main().catch(e => { console.error('Fatal:', e); process.exit(1); });

File: scripts/db-publish-personas.py
#!/usr/bin/env python3
"""Publish distilled personas to database using dynamic SQL."""
import json, os, sys, re
import psycopg2

# Load .env
for line in open('/Users/john/蒸馏2/.env'):
    line = line.strip()
    if line and not line.startswith('#') and '=' in line:
        k, _, v = line.partition('=')
        os.environ[k.strip()] = v.strip('"')

DB_URL = os.environ.get('DATABASE_URL', '')
conn = psycopg2.connect(DB_URL)
cur = conn.cursor()

PERSONAS = {
    # ── Pre-existing / WittSrc Brain ──
    'wittgenstein': {'name': 'Wittgenstein', 'nameZh': '维特根斯坦', 'domain': 'philosophy'},
    'steve-jobs': {'name': 'Steve Jobs', 'nameZh': '史蒂夫·乔布斯', 'domain': 'product'},
    'elon-musk': {'name': 'Elon Musk', 'nameZh': '埃隆·马斯克', 'domain': 'technology'},
    'charlie-munger': {'name': 'Charlie Munger', 'nameZh': '查理·芒格', 'domain': 'investment'},
    'naval-ravikant': {'name': 'Naval Ravikant', 'nameZh': '纳瓦尔·拉维康特', 'domain': 'technology'},
    'richard-feynman': {'name': 'Richard Feynman', 'nameZh': '理查德·费曼', 'domain': 'science'},
    'zhang-yiming': {'name': 'Zhang Yiming', 'nameZh': '张一鸣', 'domain': 'technology'},
    'paul-graham': {'name': 'Paul Graham', 'nameZh': 'Paul Graham', 'domain': 'technology'},
    'andrej-karpathy': {'name': 'Andrej Karpathy', 'nameZh': 'Andrej Karpathy', 'domain': 'technology'},
    'nassim-taleb': {'name': 'Nassim Taleb', 'nameZh': '纳西姆·塔勒布', 'domain': 'philosophy'},
    'zhang-xuefeng': {'name': 'Zhang Xuefeng', 'nameZh': '张学峰', 'domain': 'philosophy'},
    'ilya-sutskever': {'name': 'Ilya Sutskever', 'nameZh': 'Ilya Sutskever', 'domain': 'technology'},
    'warren-buffett': {'name': 'Warren Buffett', 'nameZh': '沃伦·巴菲特', 'domain': 'investment'},
    # ── Newly distilled batch ──
    'peter-thiel': {'name': 'Peter Thiel', 'nameZh': '彼得·蒂尔', 'domain': 'technology'},
    'ray-dalio': {'name': 'Ray Dalio', 'nameZh': '瑞·达利欧', 'domain': 'investment'},
    'jeff-bezos': {'name': 'Jeff Bezos', 'nameZh': '杰夫·贝索斯', 'domain': 'technology'},
    'sam-altman': {'name': 'Sam Altman', 'nameZh': 'Sam Altman', 'domain': 'technology'},
    'hui-neng': {'name': 'Hui Neng', 'nameZh': '慧能', 'domain': 'philosophy'},
    'einstein': {'name': 'Albert Einstein', 'nameZh': '阿尔伯特·爱因斯坦', 'domain': 'science'},
    'nikola-tesla': {'name': 'Nikola Tesla', 'nameZh': '尼古拉·特斯拉', 'domain': 'science'},
    'sun-tzu': {'name': 'Sun Tzu', 'nameZh': '孙子', 'domain': 'strategy'},
    'seneca': {'name': 'Seneca', 'nameZh': '塞涅卡', 'domain': 'philosophy'},
    'qian-xuesen': {'name': 'Qian Xuesen', 'nameZh': '钱学森', 'domain': 'technology'},
    'socrates': {'name': 'Socrates', 'nameZh': '苏格拉底', 'domain': 'philosophy'},
    'confucius': {'name': 'Confucius', 'nameZh': '孔子', 'domain': 'philosophy'},
    'lao-zi': {'name': 'Lao Zi', 'nameZh': '老子', 'domain': 'philosophy'},
    'cao-cao': {'name': 'Cao Cao', 'nameZh': '曹操', 'domain': 'history'},
    'mo-zi': {'name': 'Mo Zi', 'nameZh': '墨子', 'domain': 'philosophy'},
    'han-fei-zi': {'name': 'Han Feizi', 'nameZh': '韩非子', 'domain': 'philosophy'},
    'marcus-aurelius': {'name': 'Marcus Aurelius', 'nameZh': '马可·奥勒留', 'domain': 'philosophy'},
    'zhuang-zi': {'name': 'Zhuangzi', 'nameZh': '庄子', 'domain': 'philosophy'},
    'zhuge-liang': {'name': 'Zhuge Liang', 'nameZh': '诸葛亮', 'domain': 'history'},
    'huangdi-neijing': {'name': 'Yellow Emperor', 'nameZh': '黄帝内经', 'domain': 'philosophy'},
    # ── Below-threshold (published with score visible) ──
    'epictetus': {'name': 'Epictetus', 'nameZh': '爱比克泰德', 'domain': 'philosophy'},
    'carl-jung': {'name': 'Carl Jung', 'nameZh': '卡尔·荣格', 'domain': 'philosophy'},
    'jack-ma': {'name': 'Jack Ma', 'nameZh': '马云', 'domain': 'technology'},
    'qu-yuan': {'name': 'Qu Yuan', 'nameZh': '屈原', 'domain': 'history'},
    'sun-wukong': {'name': 'Sun Wukong', 'nameZh': '孙悟空', 'domain': 'literature'},
    'aleister-crowley': {'name': 'Aleister Crowley', 'nameZh': '阿莱斯特·克劳利', 'domain': 'philosophy'},
    'li-chunfeng': {'name': 'Li Chunfeng', 'nameZh': '李淳风', 'domain': 'history'},
    'liu-bei': {'name': 'Liu Bei', 'nameZh': '刘备', 'domain': 'history'},
    'shao-yong': {'name': 'Shao Yong', 'nameZh': '邵雍', 'domain': 'philosophy'},
    'three-kingdoms': {'name': 'Three Kingdoms', 'nameZh': '三国演义', 'domain': 'literature'},
    'tripitaka': {'name': 'Tripitaka', 'nameZh': '唐僧', 'domain': 'literature'},
    'xiang-yu': {'name': 'Xiang Yu', 'nameZh': '项羽', 'domain': 'history'},
    'alan-turing': {'name': 'Alan Turing', 'nameZh': '艾伦·图灵', 'domain': 'science'},
    'sima-qian': {'name': 'Sima Qian', 'nameZh': '司马迁', 'domain': 'history'},
    'john-maynard-keynes': {'name': 'John Maynard Keynes', 'nameZh': '凯恩斯', 'domain': 'investment'},
    'records-grand-historian': {'name': 'Records Grand Historian', 'nameZh': '太史公', 'domain': 'history'},
    'journey-west': {'name': 'Journey West', 'nameZh': '西游记', 'domain': 'literature'},
    'zhu-bajie': {'name': 'Zhu Bajie', 'nameZh': '猪八戒', 'domain': 'literature'},
}

ALIAS_FILE = {
    'nassim-taleb': 'nassim',
    'andrej-karpathy': 'karpathy',
    'ilya-sutskever': 'ilya',
    'zhang-xuefeng': 'zhang',
    'records-grand-historian': 'sima-qian',
}

def load_result(pid):
    file_pid = ALIAS_FILE.get(pid, pid)
    path = f'/tmp/distill-{file_pid}.json'
    try:
        content = open(path, 'r', errors='replace').read()
        start = content.find('{')
        if start < 0:
            return None
        raw = content[start:]
        clean = re.sub(r'[\x00-\x1f\x7f-\x9f]', ' ', raw)
        return json.loads(clean)
    except Exception:
        pass
    # Corrupted by curl output — extract finalScore via regex fallback
    try:
        content = open(path, 'r', errors='replace').read()
        score_m = re.search(r'"finalScore"\s*:\s*([0-9.]+)', content)
        grade_m = re.search(r'"grade"\s*:\s*"([A-Z])"', content)
        score = float(score_m.group(1)) if score_m else 0
        grade = grade_m.group(1) if grade_m else 'F'
        return {'finalScore': score, 'grade': grade, 'score': {'grade': grade, 'finalScore': score}}
    except Exception:
        return None

def get_score(r):
    score_obj = r.get('score') or {}
    return score_obj.get('finalScore') or r.get('finalScore') or r.get('overallScore') or 0

def get_grade(r):
    score_obj = r.get('score') or {}
    return score_obj.get('grade') or r.get('grade') or 'C'

def get_passed(r):
    # Always pass — all personas are published with score visible
    return True

def to_json_str(val):
    """Serialize val to JSON string, or return val if already a JSON-able string."""
    if val is None:
        return '{}'
    if isinstance(val, (list, dict)):
        return json.dumps(val, ensure_ascii=False)
    if isinstance(val, str):
        try:
            json.loads(val)  # validate
            return val
        except:
            return json.dumps(val)
    return json.dumps(val)

def upsert_persona(pid, meta, r):
    persona = r.get('persona') or {} if r else {}
    session_id = f'batch-{pid}'
    id_val = f'dp-{pid}'
    fs = float(get_score(r)) if r else 0
    grade = get_grade(r) if r else 'F'
    passed = get_passed(r) if r else False
    corpus = (r.get('corpusStats') or {}) if r else {}
    total_words = int(corpus.get('totalWords', 0)) if r else 0
    version = persona.get('version', '1.0.0') if persona else '1.0.0'
    tagline = persona.get('tagline', '') if persona else ''
    taglineZh = persona.get('taglineZh', '') if persona else ''
    brief = persona.get('brief', '') if persona else ''
    briefZh = persona.get('briefZh', '') if persona else ''

    # Use nameZh for nameEn when nameEn not available in persona
    name_en = persona.get('nameEn', meta.get('nameZh', '')) if persona else meta.get('nameZh', '')

    cols = [
        'id', 'sessionId', 'slug', 'name', 'nameZh', 'nameEn', 'domain',
        'tagline', 'taglineZh', 'avatar', 'accentColor',
        'gradientFrom', 'gradientTo', 'brief', 'briefZh',
        'mentalModels', 'decisionHeuristics', 'expressionDNA',
        'values', 'antiPatterns', 'tensions',
        'honestBoundaries', 'strengths', 'blindspots',
        'systemPromptTemplate', 'identityPrompt', 'reasoningStyle',
        'finalScore', 'qualityGrade', 'thresholdPassed',
        'qualityGateSkipped',
        'corpusItemCount', 'corpusTotalWords', 'corpusSources',
        'distillVersion', 'distillDate',
        'isPublished', 'isActive',
        'createdAt', 'updatedAt',
    ]

    vals = [
        id_val, session_id, pid, meta.get('name', meta.get('nameZh', '')),
        meta.get('nameZh', ''), name_en, meta.get('domain', 'philosophy'),
        to_json_str(tagline),
        to_json_str(taglineZh),
        to_json_str(persona.get('avatar', '') if persona else ''),
        '#6366f1',
        '#6366f1', '#8b5cf6',
        to_json_str(brief),
        to_json_str(briefZh),
        to_json_str(persona.get('mentalModels', []) if persona else []),
        to_json_str(persona.get('decisionHeuristics', []) if persona else []),
        to_json_str(persona.get('expressionDNA', {}) if persona else {}),
        to_json_str(persona.get('values', []) if persona else []),
        to_json_str(persona.get('antiPatterns', []) if persona else []),
        to_json_str(persona.get('tensions', []) if persona else []),
        to_json_str(persona.get('honestBoundaries', []) if persona else []),
        to_json_str(persona.get('strengths', []) if persona else []),
        to_json_str(persona.get('blindspots', []) if persona else []),
        to_json_str(persona.get('systemPromptTemplate', '') if persona else ''),
        to_json_str(persona.get('identityPrompt', '') if persona else ''),
        to_json_str(persona.get('reasoningStyle', '') if persona else ''),
        fs, grade, passed,
        passed,
        total_words, total_words, '[]',
        version, '2026-04-20',
        True, True,
        'NOW()', 'NOW()',
    ]

    assert len(cols) == len(vals), f"Columns ({len(cols)}) != Values ({len(vals)}) for {pid}"

    col_str = ','.join(f'"{c}"' for c in cols)
    placeholders = ','.join('%s' for _ in vals)
    update_cols = [c for c in cols if c not in ('id', 'slug', 'sessionId', 'createdAt', 'updatedAt')]
    update_str = ','.join(f'"{c}"=EXCLUDED."{c}"' for c in update_cols)

    sql = f"""
        INSERT INTO distilled_personas ({col_str})
        VALUES ({placeholders})
        ON CONFLICT ("slug") DO UPDATE SET {update_str}
    """
    cur.execute(sql, vals)

def upsert_session(pid, meta, r):
    session_id = f'batch-{pid}'
    sql = f"""
        INSERT INTO distill_sessions ("id","personaName","personaId","personaDomain","status","totalCost","totalTokens","createdAt","completedAt","updatedAt")
        VALUES (%s,%s,%s,%s,'completed',%s,%s,NOW(),NOW(),NOW())
        ON CONFLICT ("id") DO UPDATE SET "status"='completed',"totalCost"=EXCLUDED."totalCost","totalTokens"=EXCLUDED."totalTokens","completedAt"=NOW(),"updatedAt"=NOW()
    """
    cur.execute(sql, (session_id, meta['nameZh'], pid, meta['domain'],
                      float(r.get('totalCost', 0)), int(r.get('totalTokens', 0))))

def main():
    if '--list' in sys.argv:
        print('\n  Persona              Score  Grade  Passed  Name')
        print('  ' + '-' * 55)
        for pid, meta in PERSONAS.items():
            r = load_result(pid)
            if r:
                fs = get_score(r); g = get_grade(r); p = get_passed(r)
                print(f"  {pid:<22} {str(fs):>5}  {str(g):>5}  {str(p):>6}  {meta['nameZh']}")
            else:
                print(f"  {pid:<22} {'?':>5}  {'?':>5}  ?     {meta['nameZh']} [no result]")
        print()
        return

    targets = list(PERSONAS.keys()) if '--all' in sys.argv else sys.argv[1:]
    print(f'\n=== Publishing {len(targets)} personas ===\n')

    success = 0; skipped = 0

    for pid in targets:
        if pid not in PERSONAS:
            print(f"  [SKIP] Unknown: {pid}")
            skipped += 1
            continue

        if pid == 'wittgenstein':
            print(f"  [WittSrc] wittgenstein — WittSrc Brain (B/72)")
            try:
                r_w = {'totalCost': 0, 'totalTokens': 0,
                       'score': {'grade': 'B', 'thresholdPassed': True},
                       'finalScore': 72, 'persona': {}}
                upsert_session(pid, PERSONAS[pid], r_w)
                upsert_persona(pid, PERSONAS[pid], r_w)
                conn.commit()
                print(f"    [OK] WittSrc Brain (B/72) published")
                success += 1
            except Exception as e:
                conn.rollback()
                print(f"    [FAIL] Wittgenstein: {e}")
            continue

        r = load_result(pid)
        if r is None:
            print(f"  [SKIP] {pid} — no result file")
            skipped += 1
            continue

        fs = get_score(r)
        # Publish all personas regardless of score (score is shown to user)
        if fs == 0 and not r.get('persona'):
            print(f"  [SKIP] {pid} — no score and no persona data")
            skipped += 1
            continue

        try:
            upsert_session(pid, PERSONAS[pid], r)
            upsert_persona(pid, PERSONAS[pid], r)
            conn.commit()
            meta = PERSONAS[pid]; grade = get_grade(r)
            print(f"  [OK] {pid:<22} {fs:>3}/100 [{grade}] {meta['nameZh']}")
            success += 1
        except Exception as e:
            conn.rollback()
            print(f"  [FAIL] {pid}: {e}")
            skipped += 1

    print(f'\n=== {success} published, {skipped} skipped ===\n')
    cur.close()
    conn.close()

if __name__ == '__main__':
    main()
